File: projects/pinstripe/packages/@pinstripe/core/lib/database/union.js

import { Base } from '../base.js';
import { Registrable } from '../registrable.js';
import { serviceFactory } from '../service_factory.js';

export const Union = Base.extend().define(dsl => dsl
    .include(Registrable)
    .tap(Class => {
        const register = Class.register;
        Class.define(dsl => dsl
            .classProps({
                register(name){
                    serviceFactory(name, ({ database }) => database[name]);
                    return register.call(this, name);
                }
            })
        );
    })
    .classProps({
        get tableClasses(){
            if(!this.hasOwnProperty('_tableClasses')){
                this._tableClasses = {};
            }
            return this._tableClasses;
        }
    })
    .props({

        initialize(database, collections){
            this._database = database
            this._tables = [];
            (collections || this.constructor.tableClasses.map(tableClass => new tableClass(database))).forEach(collection => {
                if(collection instanceof Union){
                    this._tables.push(...collection._tables);
                } else {
                    this._tables.push(collection);
                }
            });
        },

        concat(collection){
            return new Union(this._database, [...this._tables, collection]);
        },

        all(options = {}){
            return this._database.run`${this._generateSelectSql(options)}`;
        },

        async forEach(fn){
            const rows = await this.all();
            for(let i = 0; i < rows.length; i++){
                const row = rows[i];
                await fn.call(row, row, i);
            }
            return this;
        },

        explain(options = {}){
            return this._generateSelectSql(options);
        },

        // __getMissing(name){
        //     console.log('__getMissing', name);
        //     return (...args) => new Union(this._database, this._tables.map(table => table[name](...args)));
        // },

        async _generateSelectSql(options = {}){
            const out = [];
            for(let i in this._tables){
                const table = this._tables[i];
                if(out.length){
                    out.push(' union all ');
                }
                out.push(table._generateSelectSql({
                    columnNames: await this._columnNames(),
                    orderBy: null,
                    limit: null
                }));
            }
            return this._database.sql`${out}`;
        },

        async _columnNames(){
            const out = [];
            for(let i in this._tables){
                const table = this._tables[i];
                Object.keys(await table.columns()).forEach(columnName => {
                    if(!out.includes(columnName)){
                        out.push(columnName)
                    }
                })
            }
            return out;
        }
    })
);

File: projects/slick/lib/slick/database/union.rb

require "slick/registry"
require "slick/database/adaptable"

class Slick::Database::Union

    class << self

        include Slick::Registry
        include Slick::Database::Adaptable

        def register(name, *args, &block)
            out = super
            Slick::Helper.register(name).define_method "call" do |*args, &block|
                database.send(self.class.name, *args, &block)
            end
            out
        end

        def table_classes
            @table_classes ||= []
        end

    end

    _define_adapter_methods :union

    deligate_to_adapter :_generate_select_sql

    attr_reader :database

    def initialize(database, collections = nil)
        @database = database
        @tables = []
        (collections || self.class.table_classes.map{|table_class| table_class.new(@database)}).each do |table|
            if table.kind_of?(Slick::Database::Union)
                @tables.concat(table.instance_variable_get(:@tables))
            elsif table.kind_of?(Slick::Database::Table)
                @tables << table
            end
        end
        @order_by_sql = []
        @skip_count = 0
    end

    def +(collection)
        Slick::Database::Union.new(@database, @tables.clone.push(collection))
    end

    def order_by(column, direction = 'asc')
        out = clone
        if @order_by_sql.count > 0
            @order_by_sql << [", ? #{direction.to_s == 'desc' ? 'desc' : 'asc'}", column]
        else
            @order_by_sql << ["? #{direction.to_s == 'desc' ? 'desc' : 'asc'}", column]
        end
        out
    end

    def paginate(page = 1, page_size = 10)
        out = clone
        out.instance_eval do
            @page = page
            @page_size = page_size
        end
        out
    end

    def clear_pagination
        out = clone
        out.instance_eval do
            @page = nil
            @page_size = nil
        end
        out
    end

    def skip(skip_count)
        out = clone
        out.instance_eval do
            @skip_count = skip_count
        end
        out
    end

    def count(options = {})
        @database.run(_generate_select_sql({ select: "count(*)" }.merge(options)), options).pop.values.pop
    end

    def all(options = {})
        @database.run(_generate_select_sql(options), options)
    end

    def first
        paginate(1, 1).all.first
    end

    def each(options = {}, &block)
        @database.run(_generate_select_sql(options), options).each(&block)
    end

    def explain(options = {})
        @database.send(:_prepare, _generate_select_sql(options))
    end

    def inspect
        all
    end

    def method_missing(*args, &block)
        out = clone
        out.instance_eval{ @tables = @tables.map{|table| table.send(*args, &block)} }
        out
    end

    private

    def _column_names
        @_column_names ||= begin
            out = []
            @tables.each do |table|
                table.columns.keys.each do |column_name|
                    out << column_name if !out.include?(table)
                end
            end
            out
        end
    end

    def _adapter
        @_adapter ||= database.send(:_adapter)
    end

end
